### ISSUE TITLE
[Security] Cannot use env in RememberMe config

### ISSUE THREAD

**Issue by zek** [symfony/symfony#36271]
**Symfony version(s) affected**: 3.4.39

**Description**  
This is working until today's 3.4.39 update. using `env` in `security.yaml` causes this error.

```
Invalid type for path "security.firewalls.main.remember_me.lifetime". Expected int, but got string.
```

**How to reproduce**  
inside of security.yaml
```yaml
lifetime: '%env(int:REMEMBER_ME_COOKIE_LIFETIME)%'
```

**Possible Solution**  
Remove these lines
https://github.com/symfony/symfony/pull/35910/files#diff-63961cbd5c56cdeb2e94e6024d58327fR148-R149

**Additional context**  
![image](https://user-images.githubusercontent.com/3463291/77936148-76a31100-7280-11ea-9383-42f704e68e0c.png)

**Comment by nicolas-grekas** [comment]
Which version were you upgrading from? Can you provide a small reproducing app that worked before and doesn't anymore please? That'd help ppl that want to debug this (could be you btw, up to?)

**Comment by zek** [comment]
> Which version were you upgrading from?
I was upgrading from 3.4.38.

> Can you provide a small reproducing app that worked before and doesn't anymore please?
OK I can try.

**Comment by zek** [comment]
Here is the repo https://github.com/zek/symfony-bug

I didn't complete rest of the reproduction. 
I've just added the configuration which causes error.

![image](https://user-images.githubusercontent.com/3463291/78154220-ff00ed80-7409-11ea-81d6-2fda1ed7e231.png)

**Comment by zek** [comment]
What I'm trying to do is 
use env variables in security.yaml remember_me config. So there is no special config actually.

![image](https://user-images.githubusercontent.com/3463291/78154413-3cfe1180-740a-11ea-93f2-ccbcccfbf4b1.png)

**Comment by wouterj** [comment]
Hmm, so I'm guessing this is what happends:

* `lifetime` is now an `integerNode()`
* For the Config component (which IIRC doesn't know anything about env variables), you're passing a string: `"%env(int:REMEMBER_ME_COOKIE_LIFETIME)%"`
* This throws an error, although if it wouldn't the DI component would sucessfully process the string into a integer before it's used by any PHP class.

So we either make Config aware of environment variables (that's probably a huge feature) or we revert the `integerNode()` changes (as you suggested).

@HeahDude am I mislooking something, or would reverting these 2 lines not result in much harm? (only a little less strict config processor)

**Comment by HeahDude** [comment]
Yes we should revert.

Also as @wouterj explained the Config component does not know about env processors...

It means that any node, excepted for `ScalarNode`, has potentially an issue with env variables which will be always read as strings by the loaders, an array could be `%env(json:file:...)%` etc.
Sounds very complex to try to handle this from any component, and probably too much to allow strings on every nodes...

**Comment by wouterj** [comment]
@zek as you were the one reporting and discovering the lines that broke this, are you maybe up for a PR to revert the change on the 3.4 branch? That would give you all credits you deserve :)

---

> It means that any node, excepted for ScalarNode, has potentially an issue with env variables which will be always read as strings by the loaders, an array could be %env(json:file:...)% etc.
> Sounds very complex to try to handle this from any component, and probably too much to allow strings on every nodes...

Yes, this is also what I was thinking about. I think it might be possible to "solve" it a little bit if Config would recognize that it is an env string and discover the env processor. There are [env processors for each type](https://symfony.com/doc/current/configuration/env_var_processors.html#built-in-environment-variable-processors). E.g. if something is `%env(int:)%`, we can be sure it'll be an integer when it is processed by DI.

This would make the Config component aware of the fullstack framework though. I'm not sure how to fix that, but this could be a way of solving it for all config nodes.

**Comment by nicolas-grekas** [comment]
4.4 is already able to know about these, but not 3.4 indeed.

**Comment by wouterj** [comment]
Oh, so in that case, the revert should only happen in 3.4?

**Comment by nicolas-grekas** [comment]
I suppose yes, can anyone confirm on 4.4?

**Comment by wouterj** [comment]
I can confirm that I no longer have this error when I update the reproducing application to Symfony 4.4 (thanks for providing a reproducer, that makes it very easy to confirm/test!)

**Comment by zek** [comment]
> @zek as you were the one reporting and discovering the lines that broke this, are you maybe up for a PR to revert the change on the 3.4 branch? That would give you all credits you deserve :)

Ok. I'll send a PR today. Thank you all @wouterj @HeahDude @nicolas-grekas  👍

**Comment by nicolas-grekas** [comment]
friendly ping @zek

**Comment by escopecz** [comment]
I can see this [was fixed for 3.4](https://github.com/symfony/symfony/blob/46c278316c7306de7e323866072e3c4422b14d7d/src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/RememberMeFactory.php#L146), but [it exists on 4.4+](https://github.com/symfony/symfony/blob/5.0/src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/RememberMeFactory.php#L154). How come it's not a problem on newer versions? Because we have exactly the same problem on Symfony 4.4 (https://github.com/mautic/mautic/issues/9011)

**Comment by nicolas-grekas** [comment]
Please don't comment on closed issues/PRs, you might get no follow up.
Open an issue instead, or a PR maybe?
